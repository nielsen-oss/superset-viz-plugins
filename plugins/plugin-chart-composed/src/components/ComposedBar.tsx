import React, { FC, RefObject, useEffect } from 'react';
import { BarProps } from 'recharts';
import { JsonObject } from '@superset-ui/core';
import { getBreakdownsOnly, BREAKDOWN_SEPARATOR } from './utils';

type ComposedBarProp = Partial<BarProps> & {
  rechartsDataKey: string;
  breakdown: string;
  barsUIPositionsRef: RefObject<JsonObject>;
};

const ComposedBar: FC<ComposedBarProp> = props => {
  const { x, y, height, width, fill, opacity, rechartsDataKey, breakdown, barsUIPositionsRef } = props;
  useEffect(() => {
    // @ts-ignore
    barsUIPositionsRef.current[`${rechartsDataKey}${BREAKDOWN_SEPARATOR}${getBreakdownsOnly(breakdown).join()}`] = {
      x,
      y,
      width,
      height,
    };
  }, [barsUIPositionsRef, breakdown, height, rechartsDataKey, width, x, y]);
  return <rect x={x} y={y} height={height} width={width} fill={fill} opacity={opacity} />;
};

export default ComposedBar;
