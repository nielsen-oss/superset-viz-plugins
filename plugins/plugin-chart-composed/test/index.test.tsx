/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import * as recharts from 'recharts';
import { supersetTheme, ThemeProvider } from '@superset-ui/core';
import { processNumbers } from '../src/components/utils';
import { ResultData } from '../src/components/types';
import ComposedChart from '../src/components/ComposedChart';
import ComposedChartPlugin from '../src';
import {
  allChatsLegendBottomBreakdowns,
  barsHorizontalLegendLeftY2AxisBreakdowns,
  barsHorizontalLegendTop,
  barsVerticalLegendRightNumFormatAllLabelsBreakdowns,
} from './__mocks__/composedProps';
import transformProps from '../src/plugin/transformProps';

jest.mock('recharts');

describe('plugin-chart-composed', () => {
  const RechartsComposedChart = jest.fn(props => <div {...props} />);
  const CartesianGrid = jest.fn(() => <div />);
  const Legend = jest.fn(() => <div />);
  const Tooltip = jest.fn(() => <div />);
  const XAxis = jest.fn(() => <div />);
  const YAxis = jest.fn(() => <div />);
  const Scatter = jest.fn(() => <div />);
  const Area = jest.fn(() => <div />);
  const Bar = jest.fn(() => <div />);
  const Line = jest.fn(() => <div />);
  beforeEach(() => {
    // Recharts still have some UNSAFE react functions that failing test
    jest.spyOn(console, 'warn').mockImplementation(() => null);

    jest.clearAllMocks();
    jest.resetModules();

    // @ts-ignore
    recharts.CartesianGrid = CartesianGrid;
    // @ts-ignore
    recharts.ComposedChart = RechartsComposedChart;
    // @ts-ignore
    recharts.Legend = Legend;
    // @ts-ignore
    recharts.Tooltip = Tooltip;
    // @ts-ignore
    recharts.XAxis = XAxis;
    // @ts-ignore
    recharts.YAxis = YAxis;
    // @ts-ignore
    recharts.Area = Area;
    // @ts-ignore
    recharts.Bar = Bar;
    // @ts-ignore
    recharts.Line = Line;
    // @ts-ignore
    recharts.Scatter = Scatter;
  });

  const getWrapper = (props: object) =>
    render(
      <ThemeProvider theme={supersetTheme}>
        {/*
       // @ts-ignore (no need emulate all props) */}
        <ComposedChart {...transformProps(props)} />
      </ThemeProvider>,
    );

  it('exists', () => {
    expect(ComposedChartPlugin).toBeDefined();
  });

  it('Chart with default props (legend top / horizontal / bars)', () => {
    getWrapper(barsHorizontalLegendTop);
    expect({
      ComposedChartProps: RechartsComposedChart.mock.calls[1],
      CartesianGridProps: CartesianGrid.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
      TooltipProps: Tooltip.mock.calls[1],
      XAxisProps: XAxis.mock.calls[1],
      YAxisProps: YAxis.mock.calls[1],
      BarProps: Bar.mock.calls[1],
    }).toMatchSnapshot();
    expect(Area).not.toHaveBeenCalled();
    expect(Line).not.toHaveBeenCalled();
    expect(Scatter).not.toHaveBeenCalled();
  });

  it('Chart with default props (legend left / horizontal / Y2Axis / bars / breakdowns)', () => {
    getWrapper(barsHorizontalLegendLeftY2AxisBreakdowns);
    expect({
      ComposedChartProps: RechartsComposedChart.mock.calls[1],
      CartesianGridProps: CartesianGrid.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
      TooltipProps: Tooltip.mock.calls[1],
      XAxisProps: XAxis.mock.calls[1],
      YAxisProps: YAxis.mock.calls[1],
      BarProps: Bar.mock.calls[1],
    }).toMatchSnapshot();
  });

  it('Chart with default props (legend right / vertical / number format / bars / all labels / breakdowns)', () => {
    getWrapper(barsVerticalLegendRightNumFormatAllLabelsBreakdowns);
    expect({
      ComposedChartProps: RechartsComposedChart.mock.calls[1],
      CartesianGridProps: CartesianGrid.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
      TooltipProps: Tooltip.mock.calls[1],
      XAxisProps: XAxis.mock.calls[1],
      YAxisProps: YAxis.mock.calls[1],
    }).toMatchSnapshot();
  });

  it('All charts with breakdowns', () => {
    getWrapper(allChatsLegendBottomBreakdowns);
    expect({
      ComposedChartProps: RechartsComposedChart.mock.calls[1],
      CartesianGridProps: CartesianGrid.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
      TooltipProps: Tooltip.mock.calls[1],
      XAxisProps: XAxis.mock.calls[1],
      YAxisProps: YAxis.mock.calls[1],
      BarProps: Bar.mock.calls[1],
      AreaProps: Area.mock.calls[1],
      LineProps: Line.mock.calls[1],
      ScatterProps: Scatter.mock.calls[1],
    }).toMatchSnapshot();
  });

  describe('processNumbers', () => {
    const mockData = ([
      { metric: -0.1114 },
      { metric: -0.1115 },
      { metric: -0.1116 },
      { metric: -0.1 },
      { metric: 0 },
      { metric: 0.1 },
      { metric: 0.1114 },
      { metric: 0.1115 },
      { metric: 0.1116 },
      { metric: 0.123 },
      { metric: 0.123456 },
      { metric: 123 },
      { metric: 123456 },
      { metric: 123456789 },
      { metric: 1234567891011121314151617181920 },
    ] as unknown) as ResultData[];
    it('non-adaptive', () => {
      expect(processNumbers(([{ metric: 123 }] as unknown) as ResultData[], ['metric'], 'SOME_FORMAT', 3)).toEqual([
        { metric: 123 },
      ]);
    });
    it('adaptive with digits for different numbers', () => {
      expect(processNumbers(mockData, ['metric'], 'SMART_NUMBER', 3)).toMatchSnapshot();
    });
    it('adaptive with digits without precision', () => {
      expect(processNumbers(mockData, ['metric'], 'SMART_NUMBER')).toMatchSnapshot();
    });
  });
});
