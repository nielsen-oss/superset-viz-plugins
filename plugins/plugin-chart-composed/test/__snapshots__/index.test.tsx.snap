// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-chart-composed All charts with breakdowns 1`] = `
Object {
  "AreaProps": Array [
    Object {
      "children": Array [
        undefined,
        false,
      ],
      "dataKey": "COUNT_DISTINCT(period)_$_Custom",
      "fill": "#9467bd",
      "isAnimationActive": undefined,
      "label": Object {
        "breakdown": "COUNT_DISTINCT(period)_$_Custom",
        "breakdowns": Array [
          "COUNT_DISTINCT(period)_$_Standard",
          "COUNT_DISTINCT(group_name)_$_Standard",
          "COUNT_DISTINCT(due_to_group)_$_Standard",
          "COUNT_DISTINCT(efficiency)_$_Standard",
          "COUNT_DISTINCT(period)_$_Custom",
          "COUNT_DISTINCT(group_name)_$_Custom",
          "COUNT_DISTINCT(due_to_group)_$_Custom",
          "COUNT_DISTINCT(efficiency)_$_Custom",
        ],
        "content": [Function],
        "currentData": Array [
          Object {
            "COUNT_DISTINCT(due_to_group)": 12,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 12,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 12,
            "COUNT_DISTINCT(efficiency)": 3,
            "COUNT_DISTINCT(efficiency)_$_Custom": 3,
            "COUNT_DISTINCT(efficiency)_$_Standard": 3,
            "COUNT_DISTINCT(group_name)": 31,
            "COUNT_DISTINCT(group_name)_$_Custom": 31,
            "COUNT_DISTINCT(group_name)_$_Standard": 31,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Variable",
            "period_type": "Custom",
            "rechartsDataKey": "Variable",
            "rechartsDataKeyUI": "Variable",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Direct Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Direct Due To Group",
            "rechartsDataKeyUI": "Direct Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Due To Group",
            "rechartsDataKeyUI": "Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 0,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 0,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 0,
            "COUNT_DISTINCT(efficiency)": 6,
            "COUNT_DISTINCT(efficiency)_$_Custom": 6,
            "COUNT_DISTINCT(efficiency)_$_Standard": 6,
            "COUNT_DISTINCT(group_name)": 3,
            "COUNT_DISTINCT(group_name)_$_Custom": 3,
            "COUNT_DISTINCT(group_name)_$_Standard": 3,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Summary Group",
            "period_type": "Standard",
            "rechartsDataKey": "Summary Group",
            "rechartsDataKeyUI": "Summary Group",
          },
        ],
        "fill": "black",
        "formatter": [Function],
        "hasOrderedBars": undefined,
        "position": "center",
      },
      "opacity": 0.8,
      "stroke": "#9467bd",
      "strokeWidth": 2,
      "type": "basis",
      "yAxisId": "left",
    },
    Object {},
  ],
  "BarProps": Array [
    Object {
      "children": Array [
        undefined,
        false,
      ],
      "dataKey": "COUNT_DISTINCT(efficiency)_$_Custom",
      "fill": "#7f7f7f",
      "isAnimationActive": undefined,
      "label": Object {
        "breakdown": "COUNT_DISTINCT(efficiency)_$_Custom",
        "breakdowns": Array [
          "COUNT_DISTINCT(period)_$_Standard",
          "COUNT_DISTINCT(group_name)_$_Standard",
          "COUNT_DISTINCT(due_to_group)_$_Standard",
          "COUNT_DISTINCT(efficiency)_$_Standard",
          "COUNT_DISTINCT(period)_$_Custom",
          "COUNT_DISTINCT(group_name)_$_Custom",
          "COUNT_DISTINCT(due_to_group)_$_Custom",
          "COUNT_DISTINCT(efficiency)_$_Custom",
        ],
        "content": [Function],
        "currentData": Array [
          Object {
            "COUNT_DISTINCT(due_to_group)": 12,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 12,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 12,
            "COUNT_DISTINCT(efficiency)": 3,
            "COUNT_DISTINCT(efficiency)_$_Custom": 3,
            "COUNT_DISTINCT(efficiency)_$_Standard": 3,
            "COUNT_DISTINCT(group_name)": 31,
            "COUNT_DISTINCT(group_name)_$_Custom": 31,
            "COUNT_DISTINCT(group_name)_$_Standard": 31,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Variable",
            "period_type": "Custom",
            "rechartsDataKey": "Variable",
            "rechartsDataKeyUI": "Variable",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Direct Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Direct Due To Group",
            "rechartsDataKeyUI": "Direct Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Due To Group",
            "rechartsDataKeyUI": "Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 0,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 0,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 0,
            "COUNT_DISTINCT(efficiency)": 6,
            "COUNT_DISTINCT(efficiency)_$_Custom": 6,
            "COUNT_DISTINCT(efficiency)_$_Standard": 6,
            "COUNT_DISTINCT(group_name)": 3,
            "COUNT_DISTINCT(group_name)_$_Custom": 3,
            "COUNT_DISTINCT(group_name)_$_Standard": 3,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Summary Group",
            "period_type": "Standard",
            "rechartsDataKey": "Summary Group",
            "rechartsDataKeyUI": "Summary Group",
          },
        ],
        "fill": "black",
        "formatter": [Function],
        "hasOrderedBars": undefined,
        "position": "center",
      },
      "opacity": 0.6,
      "stackId": "COUNT_DISTINCT(efficiency)_$_Custom",
      "yAxisId": "right",
    },
    Object {},
  ],
  "CartesianGridProps": undefined,
  "ComposedChartProps": undefined,
  "LegendProps": undefined,
  "LineProps": Array [
    Object {
      "children": Array [
        undefined,
        false,
      ],
      "dataKey": "COUNT_DISTINCT(group_name)_$_Custom",
      "isAnimationActive": undefined,
      "label": Object {
        "breakdown": "COUNT_DISTINCT(group_name)_$_Custom",
        "breakdowns": Array [
          "COUNT_DISTINCT(period)_$_Standard",
          "COUNT_DISTINCT(group_name)_$_Standard",
          "COUNT_DISTINCT(due_to_group)_$_Standard",
          "COUNT_DISTINCT(efficiency)_$_Standard",
          "COUNT_DISTINCT(period)_$_Custom",
          "COUNT_DISTINCT(group_name)_$_Custom",
          "COUNT_DISTINCT(due_to_group)_$_Custom",
          "COUNT_DISTINCT(efficiency)_$_Custom",
        ],
        "content": [Function],
        "currentData": Array [
          Object {
            "COUNT_DISTINCT(due_to_group)": 12,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 12,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 12,
            "COUNT_DISTINCT(efficiency)": 3,
            "COUNT_DISTINCT(efficiency)_$_Custom": 3,
            "COUNT_DISTINCT(efficiency)_$_Standard": 3,
            "COUNT_DISTINCT(group_name)": 31,
            "COUNT_DISTINCT(group_name)_$_Custom": 31,
            "COUNT_DISTINCT(group_name)_$_Standard": 31,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Variable",
            "period_type": "Custom",
            "rechartsDataKey": "Variable",
            "rechartsDataKeyUI": "Variable",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Direct Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Direct Due To Group",
            "rechartsDataKeyUI": "Direct Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Due To Group",
            "rechartsDataKeyUI": "Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 0,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 0,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 0,
            "COUNT_DISTINCT(efficiency)": 6,
            "COUNT_DISTINCT(efficiency)_$_Custom": 6,
            "COUNT_DISTINCT(efficiency)_$_Standard": 6,
            "COUNT_DISTINCT(group_name)": 3,
            "COUNT_DISTINCT(group_name)_$_Custom": 3,
            "COUNT_DISTINCT(group_name)_$_Standard": 3,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Summary Group",
            "period_type": "Standard",
            "rechartsDataKey": "Summary Group",
            "rechartsDataKeyUI": "Summary Group",
          },
        ],
        "fill": "black",
        "formatter": [Function],
        "hasOrderedBars": undefined,
        "position": "center",
      },
      "opacity": 0.8,
      "stroke": "#8c564b",
      "strokeWidth": 2,
      "type": "linear",
      "yAxisId": "left",
    },
    Object {},
  ],
  "ScatterProps": Array [
    Object {
      "children": Array [
        undefined,
        false,
      ],
      "dataKey": "COUNT_DISTINCT(due_to_group)_$_Custom",
      "fill": "#e377c2",
      "isAnimationActive": undefined,
      "label": Object {
        "breakdown": "COUNT_DISTINCT(due_to_group)_$_Custom",
        "breakdowns": Array [
          "COUNT_DISTINCT(period)_$_Standard",
          "COUNT_DISTINCT(group_name)_$_Standard",
          "COUNT_DISTINCT(due_to_group)_$_Standard",
          "COUNT_DISTINCT(efficiency)_$_Standard",
          "COUNT_DISTINCT(period)_$_Custom",
          "COUNT_DISTINCT(group_name)_$_Custom",
          "COUNT_DISTINCT(due_to_group)_$_Custom",
          "COUNT_DISTINCT(efficiency)_$_Custom",
        ],
        "content": [Function],
        "currentData": Array [
          Object {
            "COUNT_DISTINCT(due_to_group)": 12,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 12,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 12,
            "COUNT_DISTINCT(efficiency)": 3,
            "COUNT_DISTINCT(efficiency)_$_Custom": 3,
            "COUNT_DISTINCT(efficiency)_$_Standard": 3,
            "COUNT_DISTINCT(group_name)": 31,
            "COUNT_DISTINCT(group_name)_$_Custom": 31,
            "COUNT_DISTINCT(group_name)_$_Standard": 31,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Variable",
            "period_type": "Custom",
            "rechartsDataKey": "Variable",
            "rechartsDataKeyUI": "Variable",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Direct Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Direct Due To Group",
            "rechartsDataKeyUI": "Direct Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 6,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 7,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 6,
            "COUNT_DISTINCT(efficiency)": 2,
            "COUNT_DISTINCT(efficiency)_$_Custom": 2,
            "COUNT_DISTINCT(efficiency)_$_Standard": 2,
            "COUNT_DISTINCT(group_name)": 6,
            "COUNT_DISTINCT(group_name)_$_Custom": 7,
            "COUNT_DISTINCT(group_name)_$_Standard": 6,
            "COUNT_DISTINCT(period)": 3,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 3,
            "group_type": "Due To Group",
            "period_type": "Standard",
            "rechartsDataKey": "Due To Group",
            "rechartsDataKeyUI": "Due To Group",
          },
          Object {
            "COUNT_DISTINCT(due_to_group)": 0,
            "COUNT_DISTINCT(due_to_group)_$_Custom": 0,
            "COUNT_DISTINCT(due_to_group)_$_Standard": 0,
            "COUNT_DISTINCT(efficiency)": 6,
            "COUNT_DISTINCT(efficiency)_$_Custom": 6,
            "COUNT_DISTINCT(efficiency)_$_Standard": 6,
            "COUNT_DISTINCT(group_name)": 3,
            "COUNT_DISTINCT(group_name)_$_Custom": 3,
            "COUNT_DISTINCT(group_name)_$_Standard": 3,
            "COUNT_DISTINCT(period)": 5,
            "COUNT_DISTINCT(period)_$_Custom": 5,
            "COUNT_DISTINCT(period)_$_Standard": 5,
            "group_type": "Summary Group",
            "period_type": "Standard",
            "rechartsDataKey": "Summary Group",
            "rechartsDataKeyUI": "Summary Group",
          },
        ],
        "fill": "black",
        "formatter": [Function],
        "hasOrderedBars": undefined,
        "position": "center",
      },
      "opacity": 0.8,
      "shape": "circle",
      "yAxisId": "left",
    },
    Object {},
  ],
  "TooltipProps": undefined,
  "XAxisProps": undefined,
  "YAxisProps": Array [
    Object {
      "angle": -0,
      "axisLine": false,
      "dx": -5,
      "label": Object {
        "angle": -0,
        "position": "insideRight",
        "value": undefined,
      },
      "orientation": "right",
      "tick": [Function],
      "width": 1,
      "yAxisId": "right",
    },
    Object {},
  ],
}
`;

exports[`plugin-chart-composed Chart with default props (legend left / horizontal / Y2Axis / bars / breakdowns) 1`] = `
Object {
  "BarProps": undefined,
  "CartesianGridProps": undefined,
  "ComposedChartProps": undefined,
  "LegendProps": undefined,
  "TooltipProps": undefined,
  "XAxisProps": undefined,
  "YAxisProps": Array [
    Object {
      "angle": -0,
      "axisLine": false,
      "dx": -5,
      "label": Object {
        "angle": -0,
        "position": "insideRight",
        "value": "Label 2",
      },
      "orientation": "right",
      "tick": [Function],
      "width": 11,
      "yAxisId": "right",
    },
    Object {},
  ],
}
`;

exports[`plugin-chart-composed Chart with default props (legend right / vertical / number format / bars / all labels / breakdowns) 1`] = `
Object {
  "CartesianGridProps": undefined,
  "ComposedChartProps": undefined,
  "LegendProps": undefined,
  "TooltipProps": undefined,
  "XAxisProps": undefined,
  "YAxisProps": undefined,
}
`;

exports[`plugin-chart-composed Chart with default props (legend top / horizontal / bars) 1`] = `
Object {
  "BarProps": Array [
    Object {
      "children": Array [
        undefined,
        false,
      ],
      "dataKey": "COUNT_DISTINCT(group_name)",
      "fill": "#ff7f0e",
      "isAnimationActive": undefined,
      "label": Object {
        "breakdown": "COUNT_DISTINCT(group_name)",
        "breakdowns": Array [
          "COUNT_DISTINCT(period)",
          "COUNT_DISTINCT(group_name)",
        ],
        "content": [Function],
        "currentData": Array [
          Object {
            "COUNT_DISTINCT(group_name)": 7,
            "COUNT_DISTINCT(period)": 5,
            "group_type": "Direct Due To Group",
            "rechartsDataKey": "Direct Due To Group",
            "rechartsDataKeyUI": "Direct Due To Group",
          },
          Object {
            "COUNT_DISTINCT(group_name)": 7,
            "COUNT_DISTINCT(period)": 5,
            "group_type": "Due To Group",
            "rechartsDataKey": "Due To Group",
            "rechartsDataKeyUI": "Due To Group",
          },
          Object {
            "COUNT_DISTINCT(group_name)": 3,
            "COUNT_DISTINCT(period)": 5,
            "group_type": "Summary Group",
            "rechartsDataKey": "Summary Group",
            "rechartsDataKeyUI": "Summary Group",
          },
          Object {
            "COUNT_DISTINCT(group_name)": 31,
            "COUNT_DISTINCT(period)": 5,
            "group_type": "Variable",
            "rechartsDataKey": "Variable",
            "rechartsDataKeyUI": "Variable",
          },
        ],
        "fill": "white",
        "formatter": [Function],
        "hasOrderedBars": undefined,
        "position": "center",
      },
      "opacity": 1,
      "stackId": "COUNT_DISTINCT(group_name)",
      "yAxisId": "left",
    },
    Object {},
  ],
  "CartesianGridProps": undefined,
  "ComposedChartProps": undefined,
  "LegendProps": undefined,
  "TooltipProps": undefined,
  "XAxisProps": undefined,
  "YAxisProps": undefined,
}
`;

exports[`plugin-chart-composed processNumbers adaptive with digits for different numbers 1`] = `
Array [
  Object {
    "metric": -0.111,
  },
  Object {
    "metric": -0.112,
  },
  Object {
    "metric": -0.112,
  },
  Object {
    "metric": -0.1,
  },
  Object {
    "metric": 0,
  },
  Object {
    "metric": 0.1,
  },
  Object {
    "metric": 0.111,
  },
  Object {
    "metric": 0.112,
  },
  Object {
    "metric": 0.112,
  },
  Object {
    "metric": 0.123,
  },
  Object {
    "metric": 0.123,
  },
  Object {
    "metric": 123,
  },
  Object {
    "metric": 123456,
  },
  Object {
    "metric": 123456789,
  },
  Object {
    "metric": 1.2345678910111213e+30,
  },
]
`;

exports[`plugin-chart-composed processNumbers adaptive with digits without precision 1`] = `
Array [
  Object {
    "metric": -0.1114,
  },
  Object {
    "metric": -0.1115,
  },
  Object {
    "metric": -0.1116,
  },
  Object {
    "metric": -0.1,
  },
  Object {
    "metric": 0,
  },
  Object {
    "metric": 0.1,
  },
  Object {
    "metric": 0.1114,
  },
  Object {
    "metric": 0.1115,
  },
  Object {
    "metric": 0.1116,
  },
  Object {
    "metric": 0.123,
  },
  Object {
    "metric": 0.123456,
  },
  Object {
    "metric": 123,
  },
  Object {
    "metric": 123456,
  },
  Object {
    "metric": 123456789,
  },
  Object {
    "metric": 1.2345678910111213e+30,
  },
]
`;
